### ISSUE TITLE
Evaluate network in training=True configuration

### ISSUE THREAD

**Issue by clkruse** [tensorflow/tfjs#3152]
To get help from the community, we encourage using Stack Overflow and the [`tensorflow.js`](https://stackoverflow.com/questions/tagged/tensorflow.js) tag.

#### TensorFlow.js version: 
1.7.3

#### Browser version: Chrome 
81.0.4044.113

#### Describe the problem or feature request: 
I have trained a pix2pix network in python, and converted it to a tfjs graph model. In python, I can evaluate the network with the `training=True` flag, which is required in order for the batch normalization layers to use the sample mean and variance rather than the mean and variance from the training set. Is it possible to either call the model in a training configuration, or to modify the batchnorm layer to use sample statistics rather than training statistics?

There is a similar issue here: #562

**Comment by rthadur** [comment]
@clkruse please share your code if possible ?

**Comment by clkruse** [comment]
@rthadur There's not too much on the code front to show here.

In python, the model is defined according to the [Tensorflow Pix2Pix example](https://www.tensorflow.org/tutorials/generative/pix2pix#build_the_generator).

On the Tensorflow JS side, I converted the keras saved .h5 using `tensorflowjs_converter --input_format=keras --output_format=tfjs_graph_model --quantization_bytes=1 --weight_shard_size_bytes=4194304 generator_e90_combined.h5 combo_e90`

To evaluate the model, my simplified code is:
```
async function start(modelPath) {
    //load the model
    model = await tf.loadGraphModel(modelPath);

function predict(imgData) {
    return tf.tidy(() => {
        const gImg = model.predict(sketchTensor);
        const postImg = postprocess(gImg);
        return postImg;
    })
}
```

As a comparison in outputs, here is an example of the difference between calling a model in python with training=True and False. Note, the network here has only been trained for a few epochs, but the output from the training=True configuration is preferred:
![Combined Epoch 90 - Training True and False](https://user-images.githubusercontent.com/13071901/80559722-bd794900-899b-11ea-9e6b-788baf8b31c5.png)

If it's any help, the current tfjs demo with the issue can be found here: https://calebkruse.com/edges2legos/

**Comment by pyu10055** [comment]
@clkruse Hi, we are not familiar with the train=True option, what is the API you used for python inference? can you show your python code for that? thanks.

**Comment by clkruse** [comment]
Hi @pyu10055. Thanks for the response!

The python tensorflow code I'm referencing is in this portion of the [TF pix2pix example](https://www.tensorflow.org/tutorials/generative/pix2pix#generate_images), with this comment about the `training=True` configuration:
> Note: The training=True is intentional here since we want the batch statistics while running the model on the test dataset. If we use training=False, we will get the accumulated statistics learned from the training dataset (which we don't want) 

The code in question looks like this:
```
def generate_images(model, test_input, tar):
  prediction = model(test_input, training=True)
  plt.figure(figsize=(15,15))

  display_list = [test_input[0], tar[0], prediction[0]]
  title = ['Input Image', 'Ground Truth', 'Predicted Image']

  for i in range(3):
    plt.subplot(1, 3, i+1)
    plt.title(title[i])
    # getting the pixel values between [0, 1] to plot it.
    plt.imshow(display_list[i] * 0.5 + 0.5)
    plt.axis('off')
  plt.show()
```

**Comment by B0RJA** [comment]
I am also interested in this feature.
@clkruse Did you find a solution or workaround?

**Comment by clkruse** [comment]
@B0RJA unfortunately, the issue was never fixed and I wasn't able to create a workaround.

I believe @zaidalyafeai found this issue in #562 and created a workaround in tfjs found [here](https://github.com/zaidalyafeai/zaidalyafeai.github.io/tree/master/pix2pix), but the fix only works with older versions of TF and tfjs, and the fix never made it into tfjs.

**Comment by zaidalyafeai** [comment]
I think the fix should be simple in theory. I have checked the source code and it seems there is a `training` flag. However, not sure if this is the correct one https://github.com/tensorflow/tfjs/blob/98435f4f70e6afeec00cd3bc91a46e404aba4b1b/tfjs-layers/src/layers/normalization.ts#L331

**Comment by B0RJA** [comment]
Thanks to both of you! I will give the `training` flag a try. I've been stuck with this for days. In my case I use the pytorch implementation of [pix2pix](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/) to convert it to onnx, then to keras and finally to tensorflow.js. Everything works fine in the browser but the results of the inference are different. I though the problem was in the conversion. Recent version of onnx includes a param TrainingMode to preserve the behavior of batchnorm and dropout layers but this feature is missing in tensorflow.js.

**Comment by zaidalyafeai** [comment]
@B0RJA oh, that must be a lot of work. I remember trying to train the pix2pix model without the flag and it gave very bad results. Anyhow, If you need any help let me know.

**Comment by B0RJA** [comment]
The solution is to use the [apply](https://js.tensorflow.org/api/latest/#tf.layers.Layer.apply) method with the `training` kwarg set to `true`. Tested with TensorFlow.js 2.0.

`const output = model.apply(input, {'training': true});`

**Comment by clkruse** [comment]
Hi @B0RJA, I'm trying to make a pix2pix example again, but am running into the same issue. When you use the apply method, are you modifying individual layers in the loaded model? If I load a model using `loadGraphModel` and then try to use `model.apply(input)` it throws an error saying that `model.apply` is not a function.

Do you have any suggestions?

**Comment by B0RJA** [comment]
I used the [pytorch](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix) implementation of pix2pix, then converted the model to keras using [pytorch2keras](https://github.com/nerox8664/pytorch2keras) and finally converted it to tf using [tensorflowjs](https://pypi.org/project/tensorflowjs/). In the resulting json I had to change `"class_name": "Model"` and finally in the html:

```
var model = await tf.loadLayersModel('tfjs/model.json'); // 60MB
...
const prediction = model.apply(input, {'training': true});
const data = await prediction.data();
```

**Comment by nonoesp** [comment]
> I used the [pytorch](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix) implementation of pix2pix, then converted the model to keras using [pytorch2keras](https://github.com/nerox8664/pytorch2keras) and finally converted it to tf using [tensorflowjs](https://pypi.org/project/tensorflowjs/). In the resulting json I had to change `"class_name": "Model"` and finally in the html:
> 
> ```
> var model = await tf.loadLayersModel('tfjs/model.json');
> ...
> const prediction = model.apply(input, {'training': true});
> const data = await prediction.data();
> ```

Wow! Thanks — that worked perfectly @B0RJA. 🙃

**Comment by AmitMY** [comment]
Works for me too :) I just hope that it does not "cost" more compute to run it in training mode
